stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -0,0 +1,162 @@
+      }
+
+      /* Reverse mode differentiation 
+       * algorithm refernce: 
+       *
+       * Mike Giles. An extended collection of matrix 
+       * derivative results for forward and reverse mode AD. 
+       * Jan. 2008.
+       *
+       * Note algorithm  as laid out in Giles is 
+       * row-major, so Eigen::Matrices are explicitly storage
+       * order RowMajor, whereas Eigen defaults to 
+       * ColumnMajor. Also note algorithm 
+       * starts by calculating the adjoint for 
+       * A(M_ - 1, M_ - 1), hence pos on line 94 is decremented
+       * to start at pos = M_ * (M_ + 1) / 2.
+       * */
+      virtual void chain() {
+        using Eigen::Matrix;
+        using Eigen::Dynamic;
+        using Eigen::RowMajor;
+        Matrix<double, Dynamic, Dynamic, RowMajor> adjL(M_, M_);
+        Matrix<double, Dynamic, Dynamic, RowMajor> LA(M_, M_);
+        Matrix<double, Dynamic, Dynamic, RowMajor> adjA(M_, M_);
+        size_t pos = 0;
+        for (size_type i = 0; i < M_; ++i)
+          for (size_type j = 0; j <= i; ++j) {
+            adjL.coeffRef(i, j) = _variRefL[pos]->adj_;
+            LA.coeffRef(i, j) = _variRefL[pos]->val_;
+            ++pos;
+          }
+
+        --pos;
+        for (int i = M_ - 1; i >= 0; --i) {
+          for (int j = i; j >= 0; --j) {
+            if (i == j) {
+              adjA.coeffRef(i, j) = 0.5 * adjL.coeffRef(i, j)
+                / LA.coeffRef(i, j);
+            } else {
+              adjA.coeffRef(i, j) = adjL.coeffRef(i, j)
+                / LA.coeffRef(j, j);
+              adjL.coeffRef(j, j) = adjL.coeffRef(j, j)
+                - adjL.coeffRef(i, j) * LA.coeffRef(i, j) / LA.coeffRef(j, j);
+            }
+            for (int k = j - 1; k >=0; --k) {
+              adjL.coeffRef(i, k) = adjL.coeffRef(i, k)
+                - adjA.coeffRef(i, j) * LA.coeffRef(j, k);
+              adjL.coeffRef(j, k) = adjL.coeffRef(j, k)
+                - adjA.coeffRef(i, j) * LA.coeffRef(i, k);
+            }
+            _variRefA[pos--]->adj_ += adjA.coeffRef(i, j);
+          }
+        }
+      }
+    };
+
+    /* Reverse mode specialization of
+     * cholesky decomposition
+     *
+     * Internally calls llt rather than using 
+     * stan::math::cholesky_decompose in order
+     * to use selfadjointView<Lower> optimization.
+     *
+     * Note chainable stack varis are created
+     * below in Matrix<var, -1, -1>
+     *
+     * @param Matrix A
+     * @return L cholesky factor of A
+     */
+    Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>
+    cholesky_decompose(const Eigen::Matrix<var, -1, -1> &A) {
+      stan::math::check_square("cholesky_decompose", "A", A);
+      stan::math::check_symmetric("cholesky_decompose", "A", A);
+
+      Eigen::Matrix<double, -1, -1> L_A(value_of_rec(A));
+      Eigen::LLT<Eigen::MatrixXd> L_factor
+        = L_A.selfadjointView<Eigen::Lower>().llt();
+      check_pos_definite("cholesky_decompose", "m", L_factor);
```

Comment:
Thanks for all the clarifications.  I didn't catch
the L_factor argument!  That's a neat solution.

I've always seen those countdown loops done the way I
wrote it, but it's not big deal either way.  It's not
an efficiency issue.  Counting up might be, though.  I
found big differences profiling in various cases.

- Bob

> On Aug 4, 2015, at 3:23 PM, Marcus Brubaker <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/fun/cholesky_decompose.hpp:
> 
> > +     *
> > +     * Note chainable stack varis are created
> > +     * below in Matrix<var, -1, -1>
> > +     *
> > +     * @param Matrix A
> > +     * @return L cholesky factor of A
> > +     */
> > +    Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>
> > +    cholesky_decompose(const Eigen::Matrix<var, -1, -1> &A) {
> > +      stan::math::check_square("cholesky_decompose", "A", A);
> > +      stan::math::check_symmetric("cholesky_decompose", "A", A);
> > +
> > +      Eigen::Matrix<double, -1, -1> L_A(value_of_rec(A));
> > +      Eigen::LLT<Eigen::MatrixXd> L_factor
> > +        = L_A.selfadjointView<Eigen::Lower>().llt();
> > +      check_pos_definite("cholesky_decompose", "m", L_factor);
> 
> This actually avoids the cost of check_pos_definite as there is a specialization which works directly on the Eigen::LLT object which is what gets passed in here.
> 
> —
> Reply to this email directly or view it on GitHub.
>